### ISSUE TITLE
Misleading exception string when referencing non-DataFrame data sources

### ISSUE THREAD

**Issue by edschofield** [altair-viz/altair#875]
Consider this:
```python
cars = vega_datasets.data.cars()
alt.Chart(cars).mark_bar().encode(
    x='count()',
    y='origin'
)
```

This exception string is helpful (the origin column should be capitalized):
```
ValueError: origin encoding field is specified without a type; the type cannot be inferred because it does not match any column in the data.
```
whereas this exception string is misleading:
```python
cars.to_json('cars.json')
alt.Chart('cars.json').mark_bar().encode(
    x='count()',
    y='origin'
)
```
```
ValueError: origin encoding field is specified without a type; the type cannot be automacially inferred because the data is not specified as a pandas.DataFrame.
```
Instead of being flagged about the incorrect column name, the hapless user may then conclude that changing the code as follows would solve the problem:
```python
cars.to_json('cars.json')
alt.Chart('cars.json').mark_bar().encode(
    x='count()',
    y='origin:N'
)
```
but this confusingly gives the plot below instead:

![visualization 2](https://user-images.githubusercontent.com/78314/40289814-e2c2d3d8-5cfd-11e8-9a90-4310f66f7d17.png)

(BTW, "automacially" in the exception string also contains a typo.)

**Comment by edschofield** [comment]
Interestingly, changing the code to refer to the correct column name still doesn't give the equivalent plot to the DataFrame input:
```python
alt.Chart('cars.json').mark_bar().encode(
    x='count()',
    y='Origin:N'
)
```

This should probably be filed as a separate issue, but I don't yet understand exactly what the issue is ...

**Comment by jakevdp** [comment]
We briefly tried identifying invalid column names at the Python level in version 1.2: it did not go well. The problem is there are many different data sources available in Vega-Lite: top-level data, named data sources, transforms, etc. and it's difficult to correctly identify all of those within the Python layer without causing problems in corner cases. So, rather than trying to raise errors in the simplest cases that lead to false alarms in more complicated cases, we chose to not warn about incorrect column names in the Python side of the code.

While the error may be misleading in this simple context, the error is not incorrect: you are specifying a column name without a type, and that column name may be a correct reference to a data source available to Vega-Lite, so it would not be good for Altair to prevent it from being used.

**Comment by jakevdp** [comment]
Another thing: we cannot assume that a file like ``data.json`` that is accessible from the Python context is identical to the URL named ``data.json`` that is accessible from the renderer's context. So if we try to be clever in Altair and read the filename to check the validity of columns, it will cause problems when the renderer URL points to a different file within its context. So that kind of check is not possible: if data is provided by URL, Altair cannot make any inferences about its content.

**Comment by jakevdp** [comment]
Typo fixed in a1dc1326ef71c. Thanks for pointing it out :smile:

**Comment by edschofield** [comment]
Great answer. Thanks, Jake!
